dali/kernels/kernel_manager.h

```diff
@@ -0,0 +1,142 @@
+    }
+    return *static_cast<Kernel*>(instance.get());
+  }
+
+  template <typename Kernel, typename... Args>
+  Kernel &get() {
+    void (*deleter)(void *) = delete_kernel<Kernel>;
+    if (!instance)
+      throw std::logic_error("The kernel instance is null");
+    if (instance.get_deleter() != deleter)
+      throw std::logic_error("The kernel instance is of different type than requested");
+
+    return *static_cast<Kernel*>(instance.get());
+  }
+
+  template <typename Kernel>
+  static void delete_kernel(void *ptr) {
+    delete static_cast<Kernel*>(ptr);
+  }
+
+  explicit operator bool() const noexcept { return static_cast<bool>(instance); }
+};
+
+/// @brief Manages multiple instances of run-time typed kernels
+///
+/// KernelManager provides type erasure for kernels whose type is selected at
+/// run-time. Kernel manager also carries out mundane tasks of keeping
+/// ScratchpadAllocators and reserving memory according to requirements returned
+/// by kernel's Setup method.
+///
+/// A scratchpad allocator is created per-thread with thread indexing supported
+/// explicitly by the caller.
+class DLL_PUBLIC KernelManager {
+ public:
+  /// @brief Creates `num_instances` slots for kernels and `num_threads` scratcapads
+  void Initialize(size_t num_instances, size_t num_threads);
+
+  /// @brief Clears kernel instances and scratchpads
+  void Reset();
+
+  /// @brief Gets or creates a Kernel instance
+  template <typename Kernel, typename... ConstructorArgs>
+  Kernel &GetInstance(int instance_idx, ConstructorArgs &&...args) {
+    return instances[instance_idx].create_or_get<Kernel>(std::forward<ConstructorArgs>(args)...);
+  }
+
+  /// @brief Gets a reference to an internally maintained copy of KernelRequirements
+  KernelRequirements &GetRequirements(int index) {
+    return instances[index].requirements;
+  }
+
+  /// @brief Gets a const-reference to an internally maintained copy of KernelRequirements
+  const KernelRequirements &GetRequirements(int index) const {
+    return instances[index].requirements;
+  }
+
+  size_t NumInstances() const { return instances.size(); }
+  size_t NumThreads() const { return scratchpads.size(); }
+
+  /// @brief Gets a scratchpad allocator assigned to a given thread.
+  ScratchpadAllocator &GetScratchadAllocator(int thread_idx) {
+    return scratchpads[thread_idx];
+  }
+
+  /// @brief Calls setup on specified kernel instance.
+  template <typename Kernel, typename... InArgs>
+  KernelRequirements &Setup(int instance_idx, KernelContext &context, InArgs &&...in_args) {
+    auto &inst = instances[instance_idx];
+    return inst.requirements = inst.get<Kernel>().Setup(context, std::forward<InArgs>(in_args)...);
+  }
+
+  /// @brief Calls Run on specified kernel instance using Scratchpad for given thread.
+  template <typename Kernel, typename... OutInArgs>
+  void Run(int thread_idx, int instance_idx, KernelContext &context, OutInArgs &&...out_in_args) {
```

Comment:
I would prefer if user was able to use Kernel API directly, and maybe provide a function for the scratchpad handling like
```
prepare_scratchpad_for_context(KernelContext &ctx, int thread_idx, KernelRequirements req);
```